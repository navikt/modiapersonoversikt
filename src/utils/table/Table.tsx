import * as React from 'react';
import { ReactNode } from 'react';
import { loggError } from '../frontendLogger';

type TitleCell = string | ReactNode;
export type TitleRow = TitleCell[];
export type TableCell = string | number | undefined | ReactNode;
export type TableRow = Array<TableCell>;
export type TableRows = TableRow[];

export type TableProps = { tittelRekke: TitleRow; rows: TableRows; rowsOnClickHandlers?: Array<() => void> };

export function Table({ tittelRekke, rows, rowsOnClickHandlers }: TableProps) {
    rows.forEach((row: TableRow) => {
        if (row.length !== tittelRekke.length) {
            loggError(new Error('Ulik lengde på tittelRekke og innholdsrekke, dette bør du nok se på'));
        }
    });
    if (rowsOnClickHandlers && rowsOnClickHandlers.length != rows.length) {
        loggError(new Error('Ulik lengde på liste med onClickHandlers og antall rows'));
    }
    return (
        <table role="table">
            <thead role="rowgroup">
                <tr role="row">
                    {tittelRekke.map((tittel: TitleCell, index: number) => (
                        <th role="columnheader" key={index}>
                            {tittel}
                        </th>
                    ))}
                </tr>
            </thead>
            <tbody role="rowgroup">
                {rows.map((row: TableRow, index: number) => (
                    <tr role="row" key={index} onClick={rowsOnClickHandlers && rowsOnClickHandlers[index]}>
                        {row.map((entry, i) => (
                            <td role="cell" key={i}>
                                {entry || (entry === 0 && '0') || '\u2014'}
                            </td>
                        ))}
                    </tr>
                ))}
            </tbody>
        </table>
    );
}
