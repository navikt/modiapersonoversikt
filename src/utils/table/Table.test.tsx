import * as React from 'react';

import { mount, shallow } from 'enzyme';
import { Table, TableRows, TitleRow } from './Table';

test('lager tabell basert på input', () => {
    const header: TitleRow = ['kolonne 1', 'kolonne 2'];
    const body: TableRows = [['rad1 kolonne 1', 'rad1 kolonne 2']];

    const result = shallow(<Table tittelRekke={header} rows={body} />);
    const expectedResult = (
        <table>
            <thead>
                <tr>
                    <th>kolonne 1</th>
                    <th>kolonne 2</th>
                </tr>
            </thead>
            <tbody>
                <tr>
                    <td>rad1 kolonne 1</td>
                    <td>rad1 kolonne 2</td>
                </tr>
            </tbody>
        </table>
    );

    expect(result).toMatchElement(expectedResult);
});

test('setter på riktige click-handlere på riktige kolonner', () => {
    const header: TitleRow = ['kolonne 1', 'kolonne 2'];
    const body: TableRows = [['rad1 kolonne 1', 'rad1 kolonne 2'], ['rad2 kolonne 1', 'rad2 kolonne 2']];
    const callbackFørsteRekke = jest.fn();
    const callbackAndreRekke = jest.fn();
    const onClickHandlere = [callbackFørsteRekke, callbackAndreRekke];
    const result = mount(<Table tittelRekke={header} rows={body} rowsOnClickHandlers={onClickHandlere} />);

    result
        .find('tbody')
        .find('tr')
        .first()
        .simulate('click')
        .simulate('click');
    result
        .find('tbody')
        .find('tr')
        .last()
        .simulate('click');

    expect(callbackFørsteRekke.mock.calls.length).toBe(2);
    expect(callbackAndreRekke.mock.calls.length).toBe(1);
});
