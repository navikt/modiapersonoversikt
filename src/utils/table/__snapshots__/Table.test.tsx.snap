// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lager tabell basert p√• input 1`] = `
<table
  role="table"
>
  <thead
    role="rowgroup"
  >
    <tr
      role="row"
    >
      <th
        role="columnheader"
      >
        kolonne 1
      </th>
      <th
        role="columnheader"
      >
        kolonne 2
      </th>
    </tr>
  </thead>
  <tbody
    data-testid="table-tbody"
    role="rowgroup"
  >
    <tr
      role="row"
    >
      <td
        role="cell"
      >
        rad1 kolonne 1
      </td>
      <td
        role="cell"
      >
        rad1 kolonne 2
      </td>
    </tr>
  </tbody>
</table>
`;
